js/index.d.ts

```diff
@@ -0,0 +1,122 @@
+/// <reference lib="es2020" />
+
+declare interface MimeType {
+    mime: string;
+}
+
+export interface Resource {
+    name: string;
+    aliases: string[];
+    kind: MimeType | "template";
+    content: string;
+    dependencies?: string[];
+    permission?: number;
+}
+
+export interface FilterListMetadata {
+    homepage?: string;
+    title?: string;
+    expires?: number;
+    redirect?: string;
+}
+
+export interface ParseOptions {
+    format: typeof FilterFormat;
+    rule_types: typeof RuleTypes;
+}
+
+declare interface BlockerResult {
+    matched: boolean,
+    important: boolean,
+    redirect?: string,
+    rewritten_url?: string,
+    exception?: string,
+    filter?: string,
+}
+
+export class Engine {
+    constructor(rules: FilterSet, debug: boolean);
+    addResource(resource: Resource): boolean;
+    check(url: string, source_url: string, request_type: string, debug?: false): boolean;
```

Comment:
Also, `request_type` is enum in Rust. So, I think it should be enum here too. 

https://docs.rs/adblock/latest/adblock/request/enum.RequestType.html

Reply:
technically, although the enum isn't currently used in any public-facing Rust API. Those functions just accept string arguments. I do want to change that but it'll have to be left to a breaking change in the future.

Reply:
Maybe we can deal with union type here: 
```ts
type RequestType =
  | 'beacon'
  | 'csp'
  | 'document'
  | 'dtd'
  | 'fetch'
  | 'font'
  | 'image'
  | 'media'
  | 'object'
  | 'other'
  | 'ping'
  | 'script'
  | 'stylesheet'
  | 'subdocument'
  | 'websocket'
  | 'xlst'
  | 'xmlhttprequest';
  ```

For reference see different implementations from other adblockers:

https://github.com/AdguardTeam/tsurlfilter/blob/master/packages/tsurlfilter/src/request-type.ts#L6-L39

https://github.com/ghostery/adblocker/blob/0eb03a3f412c0764469f7561cf5f734eb4df4c66/packages/adblocker/src/request.ts#L58-L71